subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorsClassPath.kt

```diff
@@ -67,62 +87,139 @@ fun projectAccessorsClassPath(project: Project, classPath: ClassPath): Accessors
     }
 
-    operator fun plus(other: AccessorsClassPath) =
-        AccessorsClassPath(bin + other.bin, src + other.src)
-}
+    override fun execute(inputChanges: InputChangesInternal?, context: InputChangesContext): UnitOfWork.WorkResult {
+        withAsynchronousIO(project) {
+            buildAccessorsFor(
+                projectSchema,
+                classPath,
+                srcDir = sourcesOutputDir,
+                binDir = classesOutputDir
+            )
+        }
+        return UnitOfWork.WorkResult.DID_WORK
+    }
 
+    override fun getIdentity() = cacheKey
 
-internal
-fun cachedAccessorsClassPathFor(
-    project: Project,
-    cacheKeySpec: CacheKeySpec,
-    builder: (File, File) -> Unit
-): AccessorsClassPath {
-
-    val cacheDir =
-        scriptCacheOf(project)
-            .cacheDirFor(cacheKeySpec) { baseDir ->
-                builder(
-                    accessorsSourceDir(baseDir),
-                    accessorsClassesDir(baseDir)
-                )
-            }
+    override fun getDisplayName(): String = "Kotlin DSL accessors for $project"
+
+    override fun markExecutionTime(): Long = 0
+
+    override fun getExecutionHistoryStore(): Optional<ExecutionHistoryStore> = Optional.of(executionHistoryStore)
+
+    override fun validate(validationContext: UnitOfWork.WorkValidationContext) = Unit
+
+    override fun getChangingOutputs(): Iterable<String> =
+        listOf(sourcesOutputDir.absolutePath, classesOutputDir.absolutePath)
 
-    return AccessorsClassPath(
-        DefaultClassPath.of(accessorsClassesDir(cacheDir)),
-        DefaultClassPath.of(accessorsSourceDir(cacheDir))
+    override fun fingerprintAndFilterOutputSnapshots(
+        afterPreviousExecutionOutputFingerprints: ImmutableSortedMap<String, FileCollectionFingerprint>,
+        beforeExecutionOutputSnapshots: ImmutableSortedMap<String, FileSystemSnapshot>,
+        afterExecutionOutputSnapshots: ImmutableSortedMap<String, FileSystemSnapshot>,
+        hasDetectedOverlappingOutputs: Boolean
+    ): ImmutableSortedMap<String, CurrentFileCollectionFingerprint> = ImmutableSortedMap.copyOf(
+        afterExecutionOutputSnapshots.mapValues { (_, value) -> outputFingerprinter.fingerprint(ImmutableList.of(value)) }
     )
+
+    override fun snapshotOutputsBeforeExecution(): ImmutableSortedMap<String, FileSystemSnapshot> = snapshotOutputs()
+
+    override fun snapshotOutputsAfterExecution(): ImmutableSortedMap<String, FileSystemSnapshot> = snapshotOutputs()
+
+    private
+    fun snapshotOutputs(): ImmutableSortedMap<String, FileSystemSnapshot> {
+        val sourceSnapshots: List<FileSystemSnapshot> = fileCollectionSnapshotter.snapshot(fileCollectionFactory.fixed(sourcesOutputDir))
+        val classesSnapshots: List<FileSystemSnapshot> = fileCollectionSnapshotter.snapshot(fileCollectionFactory.fixed(classesOutputDir))
+        return ImmutableSortedMap.of(
+            SOURCES_OUTPUT_PROPERTY, CompositeFileSystemSnapshot.of(sourceSnapshots),
```

Comment:
💄 Could extract and dedupe the whole `CompositeFileSystemSnapshot.of(fileCollectionSnapshotter.snapshot(fileCollectionFactory.fixed(dir)))` expression.